exec/supervisor/plugin.go

```diff
@@ -0,0 +1,222 @@
+type Plugin struct {
+	// a list of active process instances
+	mx        sync.Mutex
+	instances map[string]pm.ProcessInstance
+
+	config *Config
+	Deps
+}
+
+// Deps are supervisor dependencies (standard infra plugins + process manager)
+type Deps struct {
+	pm pm.ProcessManager
+	infra.PluginDeps
+}
+
+// Init starts processes and their watchers defined by the configuration file
+func (p *Plugin) Init() error {
+	p.instances = make(map[string]pm.ProcessInstance)
+
+	if err := p.getConfig(); err != nil {
+		return err
+	}
+
+	// run watchers even if there are failed processes. They will be handled
+	// as terminated and exit immediately
+	err := p.start(p.config.Processes)
+	p.watch()
+
+	return err
+}
+
+// Close stops all remaining processes
+func (p *Plugin) Close() error {
+	p.mx.Lock()
+	defer p.mx.Unlock()
+
+	// TODO option to keep them running?
+
+	for name, instance := range p.instances {
+		if _, err := instance.StopAndWait(); err != nil {
+			p.Log.Errorf("supervisor close error: failed to stop process %s: %v", name, err)
+		}
+	}
+
+	return nil
+}
+
+// GetProcessNames returns names of all running process instances
+func (p *Plugin) GetProcessNames() (names []string) {
+	p.mx.Lock()
+	defer p.mx.Unlock()
+
+	for name := range p.instances {
+		names = append(names, name)
+	}
+	return names
+}
+
+// GetProcessByName returns an instance of given process
+func (p *Plugin) GetProcessByName(reqName string) pm.ProcessInstance {
+	p.mx.Lock()
+	defer p.mx.Unlock()
+
+	for name, instance := range p.instances {
+		if name == reqName {
+			return instance
+		}
+	}
+	return nil
+}
+
+func (p *Plugin) start(processes []Process) error {
+	pmLogManager := newPmLoggerManager()
+
+	var inError []string
+	for _, process := range processes {
+		if err := p.validate(&process); err != nil {
+			p.Log.Errorf("unable to start process %s: %v", process.Name, err)
+			inError = append(inError, process.Name)
+			continue
+		}
+		processStat := make(chan status.ProcessStatus)
+		pLogger, err := pmLogManager.newPmLogger(process.Name, process.LogFilePath)
+		if err != nil {
+			p.Log.Errorf("error preparing process logger %s: %v", process.Name, err)
+			inError = append(inError, process.Name)
+			continue
+		}
+		instance := p.pm.NewProcess(process.Name, process.BinaryPath,
+			pm.Args(process.Args...),
+			pm.Writer(pLogger, pLogger),
+			pm.Notify(processStat),
+			pm.AutoTerminate())
+		if err := instance.Start(); err != nil {
+			p.Log.Errorf("error starting process %s: %v", process.Name, err)
+			inError = append(inError, process.Name)
+			continue
+		}
+		p.instances[process.Name] = instance
+		p.Log.Debugf("process %s started", process.Name)
+	}
+
+	if len(inError) != 0 {
+		return errors.Errorf("following processes end up with error: %v", inError)
+	}
+	return nil
+}
+
+func (p *Plugin) validate(process *Process) error {
+	if process.Name == "" {
+		// so it will not be logged as empty space
+		process.Name = unnamedProcess
```

Comment:
Why not rather derive the name from `BinaryPath` field? It would allow multiple programs without name, which the current implementation doesn't support and will fail.